### PR TITLE
User account page refactor - Projects Tab

### PR DESCRIPTION
**Pull request by ksierks**
image

## Description of changes
What did you change in this pull request?  Provide a description of files changed, user interactions changed, etc.  Include how to test your changes.

## Related issue
Link to the GitHub issue this pull request addresses using the `#issuenum` format.  If it completes an issue, use `Fixes #issuenum` to automatically close the issue.

## Checklist
Things for the developer to confirm they've done before the PR should be accepted:

* [ ] CHANGELOG.md (and UPGRADING.md if necessary) updated with information for new change.
* [ ] Tests added (or description of how to test) for any new features.
* [ ] User documentation updated for UI or technical changes.

### CODE REVIEW

**Comment by ksierks** [comment]
Should not be merged until #1190 is complete.

File: src/main/webapp/resources/js/pages/user/components/UserProjectsPage.jsx [hunk 1]
```diff
@@ -1,10 +1,81 @@
 import React from "react";
+import { useParams } from "react-router-dom";
+import { notification, Space, Switch, Table, Typography } from "antd";
```
Comment (deepsidhu85) [inline src/main/webapp/resources/js/pages/user/components/UserProjectsPage.jsx:3]: `Table` not being used

File: src/main/webapp/resources/js/pages/user/components/UserProjectsPage.jsx [hunk 2]
```diff
@@ -1,10 +1,81 @@
...
+import { setBaseUrl } from "../../../utilities/url-utilities";
+import { formatDate } from "../../../utilities/date-utilities";
+import { useUpdateProjectSubscriptionMutation } from "../../../apis/projects/project-subscriptions";
+import { PagedTableProvider } from "../../../components/ant.design/PagedTable";
+import {
+  PagedTable,
+  PagedTableContext,
```
Comment (deepsidhu85) [inline src/main/webapp/resources/js/pages/user/components/UserProjectsPage.jsx:10]: `PagedTableContext` not being used

File: src/main/webapp/resources/js/pages/user/components/UserProjectsPage.jsx [hunk 3]
```diff
@@ -1,10 +1,81 @@
...
+      title: i18n("UserProjectsPage.table.emailSubscribed"),
+      dataIndex: "emailSubscribed",
+      key: "emailSubscribed",
+      render: (text, record) => <Switch defaultChecked={text} onChange={(checked) => updateSubscription(checked, record)} />,
+    },
+  ];
+
+  function updateSubscription(checked, record) {
+    updateProjectSubscription({ id: record.id, subscribe: checked })
+      .then((payload) => {
```
Comment (deepsidhu85) [inline src/main/webapp/resources/js/pages/user/components/UserProjectsPage.jsx:61]: Remove the `payload` and just replaced with `()`

**Review by deepsidhu85 — CHANGES_REQUESTED** [review]
A few more things..just removal of unused imports/code

File: src/main/java/ca/corefacility/bioinformatics/irida/model/subscription/ProjectSubscription.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+	}
+
+	@Override
+	public Long getId() {
+		return id;
+	}
+
+	@Override
+	public String getLabel() {
+		return "Project Subscription " + id;
```
[the remark this answers is not in the document]
Comment (ksierks) [reply src/main/java/ca/corefacility/bioinformatics/irida/model/subscription/ProjectSubscription.java:67]: This should be treated the same as toString(), right? I'm not aware of any localisation for toString().

File: src/main/webapp/resources/js/pages/user/components/UserProjectsPage.jsx (hunk 3 above)
Comment (ksierks) [reply src/main/webapp/resources/js/pages/user/components/UserProjectsPage.jsx]: Please see 41bc802d312d8a4ecdf45ce51a28db8d2b73bfbf.

File: src/main/webapp/resources/js/pages/user/components/UserProjectsPage.jsx (hunk 1 above)
Comment (ksierks) [reply src/main/webapp/resources/js/pages/user/components/UserProjectsPage.jsx]: Please see 41bc802d312d8a4ecdf45ce51a28db8d2b73bfbf.

File: src/main/webapp/resources/js/pages/user/components/UserProjectsPage.jsx [hunk 5]
```diff
@@ -1,10 +1,81 @@
...
+      })
+      .catch((error) => {
+        notification.error({
+          message: i18n("UserProjectsPage.notification.error"),
+        });
+      });
+  };
+
+  return (
+    <Space direction="vertical">
```
[the remark this answers is not in the document]
Comment (ksierks) [reply src/main/webapp/resources/js/pages/user/components/UserProjectsPage.jsx]: This makes me sad... Please see 3ac50541446d278a38cefc668f539a841b567168.

File: src/main/webapp/resources/js/apis/users/users.js [hunk 6]
```diff
@@ -60,8 +77,8 @@ export const usersApi = createApi({
 
 export const {
   useGetUserDetailsQuery,
+  useGetUserProjectDetailsQuery,
```
[the remark this answers is not in the document]
Comment (ksierks) [reply src/main/webapp/resources/js/apis/users/users.js]: My bad. Please see 4c58cea8badb00d447884bd48c871f993bf6e398.

File: src/main/java/ca/corefacility/bioinformatics/irida/ria/web/projects/ProjectSubscriptionsAjaxController.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+import org.springframework.web.bind.annotation.*;
+
+import ca.corefacility.bioinformatics.irida.model.subscription.ProjectSubscription;
+import ca.corefacility.bioinformatics.irida.ria.web.models.tables.TableRequest;
+import ca.corefacility.bioinformatics.irida.ria.web.models.tables.TableResponse;
+import ca.corefacility.bioinformatics.irida.ria.web.services.UIProjectSubscriptionService;
+import ca.corefacility.bioinformatics.irida.ria.web.users.dto.UserProjectDetailsModel;
+
+/**
+ * Controller for handling all AJAX requests for the {@link ProjectSubscription} UIs
```
[the remark this answers is not in the document]
Comment (ksierks) [reply src/main/java/ca/corefacility/bioinformatics/irida/ria/web/projects/ProjectSubscriptionsAjaxController.java]: Please see 6b53fe6f12ec0991acbf24bfee9ef4225d3c619b.

File: src/main/java/ca/corefacility/bioinformatics/irida/repositories/ProjectSubscriptionRepository.java [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+public interface ProjectSubscriptionRepository extends IridaJpaRepository<ProjectSubscription, Long> {
+
+	/**
+	 * Find all the {@link ProjectSubscription} for the given user.
+	 *
+	 * @param user the user associated with project subscriptions
+	 * @param page the page request
+	 * @return a page of {@link ProjectSubscription}.
+	 */
+	@Query("from ProjectSubscription ps where ps.user = :user")
```
[the remark this answers is not in the document]
Comment (ksierks) [reply src/main/java/ca/corefacility/bioinformatics/irida/repositories/ProjectSubscriptionRepository.java]: Please see 174c0ccc7fc3ccd34f89e06c2c12270b23cb1a98.

File: src/main/java/ca/corefacility/bioinformatics/irida/model/subscription/ProjectSubscription.java [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+	@Column(name = "email_subscription")
+	private boolean emailSubscription;
+
+	@CreatedDate
+	@NotNull
+	@Temporal(TemporalType.TIMESTAMP)
+	@Column(name = "created_date", updatable = false)
+	private Date createdDate;
+
+	public ProjectSubscription() {
```
[the remark this answers is not in the document]
Comment (ksierks) [reply src/main/java/ca/corefacility/bioinformatics/irida/model/subscription/ProjectSubscription.java:53]: Please see 609eb768e8c6b053451245ab9e5892229322e1f9.

File: src/main/java/ca/corefacility/bioinformatics/irida/model/enums/ProjectRole.java [hunk 10]
```diff
@@ -1,39 +1,77 @@
...
+
 /**
  *
  */
 public enum ProjectRole {
+	PROJECT_USER("PROJECT_USER", 1),
+	PROJECT_OWNER("PROJECT_OWNER", 2);
+
+	private String code;
+	private int level;
```
[the remark this answers is not in the document]
Comment (ksierks) [reply src/main/java/ca/corefacility/bioinformatics/irida/model/enums/ProjectRole.java]: Please see fe611e2b09d4f0b70f2072e9730b34e98bc3465c.

File: src/main/java/ca/corefacility/bioinformatics/irida/model/enums/ProjectRole.java [hunk 11]
```diff
@@ -1,39 +1,77 @@
...
+
+	/**
+	 * Get a role from the given string code
+	 *
+	 * @param code the string to get a role for
+	 * @return The requested ProjectRole
+	 */
+	public static ProjectRole fromString(String code) {
+		switch (code.toUpperCase()) {
+		case "PROJECT_USER":
```
[the remark this answers is not in the document]
Comment (ksierks) [reply src/main/java/ca/corefacility/bioinformatics/irida/model/enums/ProjectRole.java]: This was existing, not sure I use it. Please see d0dd7d059c4c0099de13113919883edebd2c2b1c. A switch with one case statement might as well be an if statement, lol.

File: src/main/java/ca/corefacility/bioinformatics/irida/ria/web/users/dto/UserProjectDetailsModel.java [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+		this.isEmailSubscribed = projectSubscription.isEmailSubscription();
+		this.roleName = projectRole.toString();
+		this.createdDate = projectSubscription.getCreatedDate();
+	}
+
+	public Long getProjectId() {
+		return projectId;
+	}
+
+	public void setProjectId(Long projectId) {
```
[the remark this answers is not in the document]
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/ria/web/users/dto/UserProjectDetailsModel.java]: It really is!

File: src/main/java/ca/corefacility/bioinformatics/irida/model/enums/ProjectRole.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/model/enums/ProjectRole.java]: Lol but you would still need to set the default value.

File: src/main/java/ca/corefacility/bioinformatics/irida/model/subscription/ProjectSubscription.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/model/subscription/ProjectSubscription.java:78]: Yeah it doesn't look like this is being used anyways. Let's leave it for now and in the future if it needs to be used we can revisit it

**Review by ericenns — APPROVED** [review]

**Pull request merged**